File: Swift/Leetcode.playground/Sources/24_DP/LC_646_MaximumLengthOfPairChain.swift
class Solution {
    func findLongestChain(_ pairs: [[Int]]) -> Int {
        let sortedPairs = pairs.sorted(by: { $0[0] < $1[0] })
        let n = sortedPairs.count
        var dp = Array(repeating: 1, count: n)
        var max = 1
        for i in 1..<n {
            for j in 0..<i {
                if sortedPairs[j][1] < sortedPairs[i][0] {
                    dp[i] = Swift.max(dp[i], dp[j] + 1)
                    max = Swift.max(max, dp[i])
                }
            }
        }
        return max
    }
}

File: Java/24_DP/LC_646_MaximumLengthOfPairChain.java
class Solution {
    public int findLongestChain(int[][] pairs) {
        Arrays.sort(pairs, (a, b) -> a[0] - b[0]);
        int n = pairs.length;
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        int max = 1;

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (pairs[j][1] < pairs[i][0]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                    max = Math.max(max, dp[i]);
                }
            }
        }

        return max;
    }
}

public class LC_646_MaximumLengthOfPairChain {

}
